evm/logic/block.py

```diff
@@ -4,26 +4,26 @@
 def blockhash(computation):
     block_number = computation.stack.pop(type_hint=constants.UINT256)
 
-    block_hash = computation.vm.get_ancestor_hash(block_number)
+    block_hash = computation.state.get_ancestor_hash(block_number)
 
     computation.stack.push(block_hash)
 
 
 def coinbase(computation):
-    computation.stack.push(computation.vm.block.header.coinbase)
+    computation.stack.push(computation.state.coinbase)
```

Comment:
No longer having to reach that deep through the class hierarchy to get the current block header is a very nice improvement, but nevertheless I wouldn't expect to find it under an attribute named `state`. Same goes for `state.precompiles` and `state.opcodes`, so I'm wondering if we couldn't instead pass those things (block, opcodes, precompiles) separately to `Computation.__init__()`, move `apply_computation()` there and have a configurable `_computation_class` instead of `_state_class` in the `VM`

Reply:
Hrm, initially I like this but..  here are my thoughts.

- `Computation` class represents the VM state while the VM is *running*
- `VMState` class represents the VM state while the VM is *at rest*

I think maintaining this separation is valuable.

Maybe there is a middle ground here where the `VMState` is treated as immutable and all of the mutability and state transformation code moves into the `Computation`.  That implies the following API.

- `VMState.coinbase`, `VMState.block_number`, ...
- `VMState.get_block_hash(number)`
- `VMState.apply_transaction(...) -> next_vm_state, computation_result`  <---- this is a pure function that returns the subsequent `VMState` after computation and the computation result... or something like that.
- `Computation.apply_computation(vm_state, message)`
- `Computation.apply_message(...)`
- `Computation.apply_create_message(...)`
- `Computation.opcodes`
- `Computation.precompiles`

This is a really rough sketch of what this might look like but it seems potentially promising.

Reply:
I like the idea if we can make `VMState` to be is treated as immutable by these changes.
(Before it, I was trying to figure out how to use another new immutable object to handle state transition.)
I can start to move these functions to be under `Computation`.

Reply:
> I can start to move these functions to be under `Computation`.

If you think this will work, lets try it.  

FYI, I'm still on *"sort-of-vacation"* time until next week but I'm setting aside time each day to check email and do code review to make sure I'm not holding anyone back.  Keep up the great work.